feat: serializable BusinessUnit

KrzysiekFel opened the pull request.

KrzysiekFel committed:
feat: serializable BusinessUnit

codecoach-dev commented on src/main/java/org/coding/serialization/BusinessUnit.java:
```diff
@@ -0,0 +1,69 @@
+
+    public BusinessUnit(String name, Employee headOfUnit, List<Employee> employees) {
+        if (!Position.MANAGER.equals(headOfUnit.getPosition())) {
+            throw new RuntimeException("Invalid position");
```
There could be a custom exception implemented, to prevent code duplication

KrzysiekFel replied:
Done

codecoach-dev commented on src/main/java/org/coding/serialization/BusinessUnit.java:
```diff
@@ -0,0 +1,69 @@
+
+    public void saveBusinessUnit(Path filePath) throws IOException {
+        Path parentPath = filePath.getParent();
+        System.out.println(parentPath.toString());
```
Remove unnecessary System.out.println

KrzysiekFel replied:
Done

codecoach-dev commented on src/main/java/org/coding/serialization/Employee.java:
```diff
@@ -0,0 +1,92 @@
+        if (!status.equals(Status.START)) {
+            throw new RuntimeException("Invalid Employee status");
+        }
+        LogUtils.info("deserialize Employee" + name, this.getClass().getName());
```
This should be higher, program may not reach that code

KrzysiekFel replied:
Done

codecoach-dev commented on src/main/java/org/coding/serialization/BackupService.java:
```diff
@@ -0,0 +1,43 @@
+import java.util.stream.Stream;
+
+public class BackupService {
+    public static void createBackupOfFiles(Path pathToBackupFrom) throws IOException {
```
Rename `pathToBackupFrom` to be more specific

KrzysiekFel replied:
Done

codecoach-dev commented on src/main/java/org/coding/serialization/BackupService.java:
```diff
@@ -0,0 +1,43 @@
+public class BackupService {
+    public static void createBackupOfFiles(Path pathToBackupFrom) throws IOException {
+        if (!Files.exists(pathToBackupFrom)) {
+            System.out.println("Provided path for backup does not exist.");
```
Throw an exception here instead of System.out

KrzysiekFel replied:
Done

codecoach-dev commented on src/main/java/org/coding/serialization/BackupService.java:
```diff
@@ -0,0 +1,43 @@
+            System.out.println("Provided path for backup does not exist.");
+            return;
+        }
+        if (!Files.isDirectory(pathToBackupFrom)) {
```
Instead of ending method, here you could add a helper method to cover single file case

KrzysiekFel replied:
Done

codecoach-dev commented on src/main/java/org/coding/serialization/BackupService.java:
```diff
@@ -0,0 +1,43 @@
+        }
+
+        Path backupFolder = pathToBackupFrom.toAbsolutePath().getParent()
+                .resolve(pathToBackupFrom.getFileName() + "Backup");
```
"_" could be added before "Backup" to separate file name

KrzysiekFel replied:
Done

codecoach-dev commented on src/main/java/org/coding/serialization/BackupService.java:
```diff
@@ -0,0 +1,43 @@
+            stream.forEach(sourcePath -> {
+                try {
+                    Path targetPath = backupFolder.resolve(pathToBackupFrom.relativize(sourcePath));
+                    if (Files.isDirectory(sourcePath)) {
```
Please double-check if lines 30-34 are necessary

KrzysiekFel replied:
They were not necessary

codecoach-dev commented on src/main/java/org/coding/serialization/BackupService.java:
```diff
@@ -0,0 +1,43 @@
+                        Files.copy(sourcePath, targetPath, StandardCopyOption.REPLACE_EXISTING);
+                    }
+                } catch (IOException e) {
+                    throw new RuntimeException(e);
```
Use our custom logger instead. LogUtils.error()

KrzysiekFel replied:
Done

codecoach-dev commented on src/test/java/serialization/BusinessUnitTest.java:
```diff
@@ -0,0 +1,94 @@
+    }
+
+    @Test
+    void creatingEmployeeSetStatusToNew() {
```
Try to use inifinite form, this could be changed to: `createEmployeeAndSetStatusToNew`

KrzysiekFel replied:
Done

codecoach-dev commented on src/test/java/serialization/BusinessUnitTest.java:
```diff
@@ -0,0 +1,94 @@
+        bu = new BusinessUnit("Product", tom, productUnitEmployees);
+    }
+
+    @Test
```
Try to re-write it with @MethodSource

KrzysiekFel replied:
Done

codecoach-dev commented on src/test/java/serialization/BusinessUnitTest.java:
```diff
@@ -0,0 +1,94 @@
+    @Test
+    void readObjectChangesEmployeeStatusToFinished() throws IOException, ClassNotFoundException {
+        Path tempFile = Files.createTempFile("businessUnit", ".ser");
+        bu.saveBusinessUnit(tempFile);
```
Create a file in test resources that will represent serialized business unit

KrzysiekFel committed:
fix: following updates were made:
    - adding specific Exceptions
    - modify test case to use ser file
    - update test case for parametrization
    - delete println
    - rename some variables
    - add error log
